-added support for iPhone X statusBar Height

SamehSalama opened the pull request:
added support for iPhone X statusBar Height

SamehSalama committed:
-added support for iPhone X statusBar Height

ppth0608 commented on BPStatusBarAlert/Classes/BPStatusBarAlert.swift:
```diff
@@ -7,6 +7,7 @@
 //
 
 import Foundation
+import UIKit
```
if using `import Foundation`, `import Foundation` do not need anymore.
I'd like to remove it :)

ppth0608 commented on BPStatusBarAlert/Classes/BPStatusBarAlert.swift:
```diff
@@ -20,6 +21,7 @@ public class BPStatusBarAlert: UIView {
         
     fileprivate var containerWindow: UIWindow?
     
+    fileprivate var deviceStatusBarHeight:CGFloat!
```
How about not using forced unwrapping, what do you think of using **computed property**
ex)
```Swift
fileprivate var deviceStatusBarHeight: CGFloat {
    if UIDevice().userInterfaceIdiom == .phone && UIScreen.main.nativeBounds.height == 2436 {
        return 15
    }
    return 0
    }
```

ppth0608 commented on BPStatusBarAlert/Classes/BPStatusBarAlert.swift:
```diff
@@ -29,12 +31,26 @@ public class BPStatusBarAlert: UIView {
     fileprivate var messageColor: UIColor = UIColor.white
     
-    fileprivate let statusBarHeight = UIApplication.shared.statusBarFrame.size.height
+    fileprivate var statusBarHeight:CGFloat!
```
This line also replace computed property is more better ^_^
```Swift
fileprivate var statusBarHeight: CGFloat {
    return UIApplication.shared.statusBarFrame.size.height + deviceStatusBarHeight
}
```

ppth0608 commented on BPStatusBarAlert/Classes/BPStatusBarAlert.swift:
```diff
@@ -29,12 +31,26 @@ public class BPStatusBarAlert: UIView {
     
     public init(duration: TimeInterval = 0.3, delay: TimeInterval = 2, position: AlertPosition = .statusBar) {
+        
+        if UIDevice().userInterfaceIdiom == .phone {
```
If using computed property, this if statement can be removed

ppth0608 commented on BPStatusBarAlert/Classes/BPStatusBarAlert.swift:
```diff
@@ -29,12 +31,26 @@ public class BPStatusBarAlert: UIView {
+            }
+        }
+        
+        //deviceStatusBarHeight = 20
```
please remove comment please^^~~

ppth0608 reviewed (changes requested):
Please check review :)
Thank you for contributing ^^

SamehSalama committed:
-removed import Foundation
-deviceStatusBarHeight and statusBarHeight are now computed properteis
-removed unnecessary comment

SamehSalama commented:
-removed import Foundation
-deviceStatusBarHeight and statusBarHeight are now computed properties
-removed unnecessary comment

hadynz commented on BPStatusBarAlert/Classes/BPStatusBarAlert.swift:
```diff
@@ -20,6 +21,14 @@ public class BPStatusBarAlert: UIView {
         
     fileprivate var containerWindow: UIWindow?
     
+    fileprivate var deviceStatusBarHeight:CGFloat {
```
There are a number of aspects to consider here:

1. Magic numbers or strings (in this case `2436`, `15` and `0`) are generally what we call [an "anti-pattern"][1] and should always be avoided to be embedded in code.
2. An argument can be made that the logic to calculate the height of the status bar is independent of this UIView class that is all about rendering. Conflating responsibilities also makes the code harder to test.

A suggestion to consider is to create another class that encapsulates this logic and [using composition][2] you'd use in this getter function.

```swift
// Create a small class with a "single responsibility" for calculating anything 
// related to a device's dimensions. You can test this class easily in isolation
class DeviceDimensions {
  var iPhoneXDeviceHeight = 2436;
  var notchHeight = 15;
  
  var device: UIDevice;
  var screen: UIScreen;

  // Pass UIDevice and UIScreen instances by reference so you can mock them in tests
  init(device: UIDevice, screen: UIScreen) {
    self.device = device;
    self.screen = screen;
  }

  var statusBarHeight: (CGFloat) {
    get {
      var hasNotch = self.device.userInterfaceIdiom == .phone && self.screen.main.nativeBounds.height == iPhoneXDeviceHeight;
      return hasNotch ? notchHeight : 0;
    }
  }
}

// Using composition, use your new class
public class BPStatusBarAlert: UIView {
  fileprivate var deviceStatusBarHeight:CGFloat {
    var deviceDimensions = new DeviceDimensions(UIDevice(), UIScreen);
    return deviceDimensions.statusBarHeight;
  }

  // ...
}
```

[1]: http://sahandsaba.com/nine-anti-patterns-every-programmer-should-be-aware-of-with-examples.html#magic-numbers-and-strings
[2]: https://www.youtube.com/watch?v=wfMtDGfHWpA

ppth0608 replied:
@hadynz  Thanks for the good comments.
I read the contents and it is a great help to me.
In particular, the DeviceDimensions class makes the constructor easier to test.
@SamehSalama,  Could you push code accepting @hadynz 's comment?

SamehSalama replied:
@hadynz   Thank you very much for the great article
@ppth0608 for sure, I will accept @hadynz 's comment, however, I would change a little bit, as for DeviceDimensions class, there is no need to add initialization method that accepts parameters of type UIDevice and UIScreen, both are singleton instances, I'd rather let them be computed properties in DeviceDimensions class and make statusBarHeight the only instance property of this class.

ppth0608 replied:
@SamehSalama Thank you for accepting comment :) And i think updated code is more Awesome than before ^^

However, it seems that some modifications are needed in the current code. (ex. Code Conventions) I will review the code soon. If only this couple is applied, it seems to be possible to merge. Thank you very much for opening PR again.

SamehSalama committed:
-added DeviceDimensions class to avoid "anti-pattern" as recommended by @hadynz

ppth0608 commented on BPStatusBarAlert/Classes/DeviceDimensions.swift:
```diff
@@ -0,0 +1,29 @@
+
+import UIKit
+
+public class DeviceDimensions {
```
it seemed to be more swift when change `class` to `struct` 

And, It would be better to assign this class's accessor to the internal level.

SamehSalama replied:
-changed to ```struct```
-It would be better to assign this class's accessor to the internal level, *would you please explain more*

ppth0608 commented on BPStatusBarAlert/Classes/DeviceDimensions.swift:
```diff
@@ -0,0 +1,29 @@
+
+public class DeviceDimensions {
+    
+    fileprivate var iPhoneXDeviceHeight:CGFloat = 2436
```
- iPhoneXDeviceHeight is constant value, so `let` is more suitable this case
- My preferred code convention is: `fileprivate let iPhoneXDeviceHeight: CGFloat = 2436`

SamehSalama replied:
done

ppth0608 commented on BPStatusBarAlert/Classes/DeviceDimensions.swift:
```diff
@@ -0,0 +1,29 @@
+public class DeviceDimensions {
+    
+    fileprivate var iPhoneXDeviceHeight:CGFloat = 2436
+    fileprivate var paddingHeight:CGFloat = 15
```
`var` -> `let`

SamehSalama replied:
done

ppth0608 commented on BPStatusBarAlert/Classes/DeviceDimensions.swift:
```diff
@@ -0,0 +1,29 @@
+        return UIScreen.main
+    }
+    
+    var statusBarHeight:CGFloat {
```
It seems likely to be omitted.
```Swift
var statusBarHeight: CGFloat {
    let  hasNotch = device.userInterfaceIdiom == .phone && screen.nativeBounds.height == iPhoneXDeviceHeight
    return hasNotch ? paddingHeight : 0
}
```

SamehSalama replied:
done :)

ppth0608 commented on BPStatusBarAlert/Classes/DeviceDimensions.swift:
```diff
@@ -0,0 +1,29 @@
+    
+    var statusBarHeight:CGFloat {
+        get {
+            let hasNotch = self.device.userInterfaceIdiom == .phone && self.screen.nativeBounds.height == self.iPhoneXDeviceHeight
```
could you remove `self` keyword? :)

SamehSalama replied:
done

ppth0608 reviewed (changes requested):
Thank you for your great code :)
But i want change there little code cuz my convention ^^

SamehSalama committed:
-DeviceDimensions is now a struct
-constants are defined using let keyword as they won't change
-removed self keywords

SamehSalama committed:
-changed DeviceDimensions class's accessor to the internal level

ppth0608 reviewed (approved):
👏🙌🏻
Thank you for contribute :)
You're great developer ^^
